src/game/gui/text_render.c

```diff
@@ -304,39 +375,162 @@ static void text_render_len(const text_settings *settings, text_mode mode, int x
+            shadow_ptr++;
+
+            // Render to the right direction
+            if(settings->direction == TEXT_HORIZONTAL) {
+                mx += w + settings->cspacing;
+            } else {
+                my += char_h;
+            }
+        }
+
+        // Render regular color characters.
+        mx = mxstart;
+        my = mystart;
+        while((*ptr != 0) && (ptr < line_end)) {
+            // Skip line endings.
+            if(*ptr == '\n') {
+                ptr++;
                 continue;
             }
 
-            render_char_shadow_surface(settings, sur, mx + start_x, my + start_y);
-            render_char_surface(settings, mode, sur, mx + start_x, my + start_y);
+            // Render character
+            w = text_render_char(cached_text, settings, mode, mx + start_x, my + start_y, *ptr, false);
+            ptr++;
 
             // Render to the right direction
             if(settings->direction == TEXT_HORIZONTAL) {
-                mx += sur->w + settings->cspacing;
+                mx += w + settings->cspacing;
             } else {
                 my += char_h;
             }
         }
 
-        ptr += advance;
+        ptr = prevptr + line_len;
         line++;
     }
 }
 
-void text_render(const text_settings *settings, text_mode mode, int x, int y, int w, int h, const char *text) {
-    text_render_len(settings, mode, x, y, w, h, text, strlen(text));
+void text_objects_free(text_object *cached_text, size_t count) {
+    for(size_t i = 0; i < count; i += 1) {
+        omf_free(cached_text[i].letters);
+    }
 }
 
-void text_render_str(const text_settings *settings, text_mode mode, int x, int y, int w, int h, const str *text) {
-    text_render_len(settings, mode, x, y, w, h, str_c(text), str_size(text));
+void text_objects_invalidate(text_object *text_cache, size_t cache_size) {
+    for(size_t i = 0; i < cache_size; i++) {
+        text_cache[i].dirty = true;
+    }
 }
+
+void text_render(text_object *cached_text, const text_settings *settings, text_mode mode, int x, int y, int w, int h,
+                 const char *text) {
+    bool dirty = false;
+    dirty |= (cached_text->x != x);
+    dirty |= (cached_text->y != y);
+    if(w > 0) {
+        dirty |= (cached_text->w != w);
+    }
+    if(h > 0) {
+        dirty |= (cached_text->h != h);
+    }
+    // dirty |= (cached_text->text != text);
+    if(cached_text->text) {
+        dirty |= (cached_text->text != text);
+        // dirty |= strncmp(cached_text->text, text, min2(strlen(cached_text->text), strlen(text))) != 0;
+    }
+    dirty |= (cached_text->mode != mode);
+    dirty |= (memcmp(&cached_text->settings, settings, sizeof(cached_text->settings)) != 0);
+
+    if(cached_text->dirty || dirty) {
+        size_t text_len = strlen(text);
+        if(text_len == 0) {
+            return;
+        }
+
+        if(cached_text->max_letters < (text_len * 4)) {
+            size_t size = (text_len * sizeof(letter) * 4);
```

Comment:
What is magic number 4?

Reply:
It looks like 4 is the number of draw calls per letter.
At the very least, it should be a named constant.

Perhaps all four draw calls should be stored in one `struct letter`? This would increase the number of fields in `struct letter`, but should decrease overall memory usage.

Reply:
It's the shadows, we support up to 3 shadows per letter, including the letter itself. Actually per the code we support 4 shadows per letter (+letter). But I've never seen more than 3 being used.